routers/api/v1/repo/pull.go

```diff
@@ -1085,8 +1085,8 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
 	log.Info("Base branch: %s", baseBranch)
 	log.Info("Repo path: %s", ctx.Repo.GitRepo.Path)
 	// Check if base branch is valid.
-	if !ctx.Repo.GitRepo.IsBranchExist(baseBranch) {
-		ctx.NotFound("IsBranchExist")
+	if !ctx.Repo.GitRepo.IsBranchExist(baseBranch) && !ctx.Repo.GitRepo.IsTagExist(baseBranch) {
```

Comment:
Why a "branch name" should be checked against "tag"?

Reply:
@wxiaoguang 

Because the Compare API was supported in https://github.com/go-gitea/gitea/pull/30349, and this API will bring in tag formats such as v1.0.0...v1.1.0, it is necessary to further determine if it is a tag.

Reply:
Then there are still questions:

* Is it the desired behavior? For example, the same as GitHub, or well-documented.
* What happens if a branch and a tag have the same name?

Reply:
> Is it the desired behavior? For example, the same as GitHub, or well-documented.

Currently, only branch names are supported, which differs from the design of the Web Router. This API was originally designed only for Pull Requests. I just added support for Tag names.

> What happens if a branch and a tag have the same name?

It won't be affected by branch names or tag names with the same name. Because the final function used supports both branch names and tag names.

https://github.com/go-gitea/gitea/blob/e865de1e9d65dc09797d165a51c8e705d2a86030/routers/api/v1/repo/pull.go#L1158-L1163

```go
	compareInfo, err := headGitRepo.GetCompareInfo(repo_model.RepoPath(baseRepo.Owner.Name, baseRepo.Name), baseBranch, headBranch, false, false)
	if err != nil {
		headGitRepo.Close()
		ctx.Error(http.StatusInternalServerError, "GetCompareInfo", err)
		return nil, nil, nil, nil, "", ""
	}
```

And you can reference the compare function in the web router.

https://github.com/go-gitea/gitea/blob/e865de1e9d65dc09797d165a51c8e705d2a86030/routers/web/repo/compare.go#L552-L556

Reply:
My gut tells me that this is somewhat unreasonable.

I can see most other calls to `GetCompareInfo` use clearly defined "refs/xxx" prefix, but here the code is just making the function to guess what the ref type it would be.

For the "web routes", I can see these code is quite old, at that time there were some design problems.

But for new code, I believe each git ref should clearly have its type, branch is branch, tag is tag, I guess we should never store a "tag" value into a variable called "branch" anymore, it is really confusing and would case more problems in the future.

----

So let's take a look at this problem again, I could understand it is going to add "tag comparison" support for the API, but the question is: is it possible to define the behavior clearly instead of making git to guess? And avoiding use "branch" variables to store "tag" values? For example: let callers pass the name with "refs/tags" prefix?

Reply:
Ideally, the logic could be like this:

```
if base has prefix "refs/heads/" OR base has prefix "refs/tags/" {
    baseRef = base
} else if IsBranchExist() {
    baseRef = "refs/heads/" + base
} else if IsTagExist() {
    baseRef = "refs/tags/" + base
}
```

It also matches what GitHub does: https://github.com/octocat/linguist/compare/v2.2.0...octocat:refs/tags/v2.3.3

If there is a tag having the same name as a branch, end users could still use `refs/tags/name` to refer it.